crates/router/src/connector/utils.rs

```diff
@@ -525,3 +526,28 @@ pub fn collect_and_sort_values_by_removing_signature(
     values.sort();
     values
 }
+
+pub fn convert_query_params_to_struct<T>(
+    query_params: Option<String>,
+) -> Result<T, error_stack::Report<errors::ConnectorError>>
+where
+    T: serde::de::DeserializeOwned,
+{
+    query_params
+        .map(|query_str| {
+            let qp: HashMap<String, String> = url::form_urlencoded::parse(query_str.as_bytes())
+                .into_owned()
+                .collect();
+            let json = Encode::<HashMap<String, String>>::encode_to_string_of_json(&qp).switch()?;
+            json.as_bytes()
+                .parse_struct(std::any::type_name::<T>())
+                .switch()
+        })
+        .transpose()?
+        .ok_or_else(|| {
+            errors::ConnectorError::MissingRequiredField {
+                field_name: "query_params",
+            }
+            .into()
+        })
```

Comment:
No the idea is to throw `MissingRequiredField ` when query_params is not present. switch() will only be called when there is a parsing error during conversion. so it won't always be `Parsing Error`. it can either be `MissingRequiredField` or `Parsing Error`. your suggestion is not handling both.

Reply:
As per my understanding we should use `MissingRequiredField ` only when necessary param missing from the merchant. To handle decoding errors for webhook body we have `errors::ConnectorError::WebhookBodyDecodingFailed`

Typo in my previous comment as switch will always return `errors::ConnectorError::ParsingFailed` cause both the methods `encode_to_string_of_json ` and `parse_struct ` will throw `ParsingError` which then will be changed to `errors::ConnectorError::ParsingFailed `. 

As it is in utils you can return `errors::ConnectorError::ParsingFailed ` as generic error and change the context in the caller methods based on your need.